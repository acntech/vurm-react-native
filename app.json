{
  "expo": {
    "name": "Vurm",
    "slug": "Vurm",
    "version": "1.0.0",
    "orientation": "portrait",
    "icon": "./assets/icon.png",
    "splash": {
      "image": "./assets/vurmLogo.png",
      "resizeMode": "contain",
      "backgroundColor": "#ffffff"
    },
    "updates": {
      "fallbackToCacheTimeout": 0
    },
    "assetBundlePatterns": [
      "**/*"
    ],
    "ios": {
      "supportsTablet": true,
      "bundleIdentifier": "com.chrisbso.Vurm"
    },
    "android": {
      "adaptiveIcon": {
        "foregroundImage": "./assets/adaptive-icon.png",
        "backgroundColor": "#FFFFFF"
      },
      "package": "com.chrisbso.Vurm"
    },
    "web": {
      "favicon": "./assets/favicon.png"
    },
    "facebookScheme": "fb498173242097723",
    "facebookAppId": "498173242097723",
    "facebookDisplayName": "Vurm",
    "facebookAdvertiserIDCollectionEnabled": false,
    "facebookAutoLogAppEventsEnabled": false
  }
}
