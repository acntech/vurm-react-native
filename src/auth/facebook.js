// Import the functions you need from the SDKs you need
import { auth } from "./firebase";
import { FacebookAuthProvider, signInWithCredential } from "firebase/auth";
import * as Facebook from "expo-facebook";

async function loginWithFacebook(setUser) {
  const appId = "498173242097723";
  await Facebook.initializeAsync({ appId });

  const { type, token } = await Facebook.logInWithReadPermissionsAsync({
    permissions: ["public_profile"],
  });

  if (type === "success") {
    // Build Firebase credential with the Facebook access token.
    const credential = FacebookAuthProvider.credential(token);

    // Sign in with credential from the Facebook user.
    signInWithCredential(auth, credential)
      .then(() => setUser(auth.currentUser))
      .catch((error) => {
        console.log(error);
      });
  }
  // return auth;
}

export default loginWithFacebook;
