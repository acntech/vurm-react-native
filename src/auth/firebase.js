import { initializeApp } from "firebase/app";
import { getAuth, onAuthStateChanged } from "firebase/auth";

const firebaseConfig = {
  apiKey: "AIzaSyB4i87NzfL1Pk_V2UpD5FSe1vwWV3-LBNs",
  authDomain: "leaderboard-rn.firebaseapp.com",
  databaseURL: "https://leaderboard-rn-default-rtdb.firebaseio.com",
  projectId: "leaderboard-rn",
  storageBucket: "leaderboard-rn.appspot.com",
  messagingSenderId: "1014828410467",
  appId: "1:1014828410467:web:35d753c749c30ee2e77a56",
  measurementId: "G-8D8MHJ3FZT",
};
//Initialize Firebase Application with our config
initializeApp(firebaseConfig);

// Listen for authentication state to change.
onAuthStateChanged(getAuth(), (user) => {
  if (user != null) {
    console.log("We are authenticated now!");
  }
});

export const auth = getAuth();
