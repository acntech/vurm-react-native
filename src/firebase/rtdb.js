import { getDatabase, ref, set, get, remove, child } from "firebase/database";

const getReference = (path) => {
  try {
    const db = getDatabase();
    return ref(db, path);
  } catch (error) {
    console.log(error);
  }
};

export const getUsersReference = () => {
  return getReference("users/");
};

export const getUserReference = (user) => {
  return getReference("users/" + user.uid);
};

export const getUserData = (user) => {
  return get(getUserReference(user));
};

export const deleteUserData = (user) => {
  if (user != null) {
    remove(getUserReference(user));
  }
};

export const setUserData = async (user, data) => {
  if (user != null) {
    const userReference = getUserReference(user);
    await set(userReference, data, (error) => {
      if (error) {
        console.log(error);
      }
    });
  }
};

export const getUserProperty = async (
  user,
  propertyName,
  defaultValue,
  setCallback
) => {
  let property = null;
  if (user != null) {
    const db = getDatabase();
    const reference = ref(db, "users/");
    await get(child(reference, user.uid))
      .then((snapshot) => {
        if (snapshot.exists()) {
          property = snapshot.val()[propertyName];
        } else {
          property = defaultValue;
        }
      })
      .catch((error) => {
        console.log(error);
      });
  }
  setCallback(property);
  return property;
};

export const extractDataFromUsersSnapshot = (usersSnapshot, userUid, topN) => {
  console.log("extract");
  let processedData = [];
  if (!usersSnapshot.hasChildren()) {
    return processedData;
  }
  let negRank = 0;
  let idx = 0;
  let prevScore = 0;
  let adjustedTopN = Object.keys(usersSnapshot.val()).length - topN - 1;

  usersSnapshot.forEach((child) => {
    const childValue = child.val();
    if (prevScore < childValue?.score) {
      negRank--;
      prevScore = childValue.score;
    }
    const youIndicator = child?.key == userUid ? " (You)" : "";
    childData = [
      negRank,
      childValue?.name + youIndicator,
      childValue?.score,
      child?.key,
    ];
    if (idx > adjustedTopN || child?.key == userUid) {
      processedData.push(childData);
    }
    idx++;
  });

  if (processedData.length > 0) {
    negRankIdx = 0;
    minNegRank = processedData[processedData.length - 1][negRankIdx];
    for (let i = 0; i < processedData.length; i++) {
      processedData[i][negRankIdx] += Math.abs(minNegRank) + 1;
    }
  }
  return processedData.reverse();
};
