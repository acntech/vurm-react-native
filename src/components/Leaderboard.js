import { useState, useEffect } from "react";
import { limitToLast, onValue, orderByChild, query } from "firebase/database";
import { getUsersReference } from "../firebase/rtdb";
import ScoreTable from "./ScoreTable";
import { auth } from "../firebase/auth";

export default Leaderboard = () => {
  const [data, setData] = useState();
  const usersReferenceOrderedByScore = query(
    getUsersReference(),
    orderByChild("score"),
    limitToLast(50)
  );

  useEffect(() => {
    const unsubscribe = onValue(
      usersReferenceOrderedByScore,
      (usersSnapshot) => {
        const leaderboardData = extractLeaderboardDataFromUsersSnapshot(
          usersSnapshot,
          null
        );
        setData(leaderboardData.reverse());
      }
    );
    return unsubscribe;
  }, []);
  return (
    <ScoreTable
      scoreData={data}
      scrollEnabled={true}
      highlightUid={auth?.currentUser?.uid}
    ></ScoreTable>
  );
};

export const extractLeaderboardDataFromUsersSnapshot = (
  usersSnapshot,
  uidFilter
) => {
  let processedData = [];
  if (!usersSnapshot.hasChildren()) {
    return processedData;
  }
  let rank = Object.keys(usersSnapshot.val()).length;
  usersSnapshot.forEach((child) => {
    const childValue = child.val();
    childData = [rank--, childValue?.name, childValue?.score, child.key];
    if (!uidFilter || uidFilter === child.key) {
      processedData.push(childData);
    }
  });
  return processedData;
};
