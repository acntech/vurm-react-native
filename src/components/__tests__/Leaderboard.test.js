import React from "react";
import { render } from "@testing-library/react-native";
import Leaderboard from "../Leaderboard";
import { onValue, query } from "firebase/database";
import { deleteApp, getApps } from "firebase/app";
import { onAuthStateChanged } from "firebase/auth";

jest.mock("firebase/database", () => ({
  ...jest.requireActual("firebase/database"),
  onValue: jest.fn(),
  query: jest.fn(),
  extractDataFromUsersSnapshot: jest.fn(),
}));

jest.mock("firebase/auth", () => ({
  ...jest.requireActual("firebase/auth"),
  onAuthStateChanged: jest.fn(),
}));

afterEach(async () => {
  apps = getApps();
  await Promise.all(apps.map((app) => deleteApp(app)));
});

describe("<Leaderboard />", () => {
  it("renders correctly", () => {
    // Arrange & Act
    children = [
      { key: "uid1", val: () => ({ score: 1, name: "BigBlueDonkey" }) },
      { key: "uid2", val: () => ({ score: 2, name: "BigRedDonkey" }) },
      { key: "uid3", val: () => ({ score: 3, name: "BigPurpleDonkey" }) },
    ];
    const snapshot = {
      val: () => children,
      forEach: (callback) => children.forEach((child) => callback(child)),
      hasChildren: () => !!children.length,
    };

    onValue.mockImplementation((ref, callback) => {
      callback(snapshot);
      return jest.fn();
    });
    query.mockImplementation((query, ...queryConstraints) => jest.fn());

    onAuthStateChanged.mockImplementation((auth, callback) => {
      return jest.fn();
    });

    const tree = render(<Leaderboard />).toJSON();

    // Assert
    expect(tree).toMatchSnapshot();
  });
});
