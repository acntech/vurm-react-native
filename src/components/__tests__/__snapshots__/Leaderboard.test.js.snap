// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Leaderboard /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Array [
          undefined,
          Object {
            "borderBottomWidth": 0,
            "borderColor": "#000",
            "borderLeftWidth": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            undefined,
            0,
            Object {
              "flexDirection": "row",
              "overflow": "hidden",
            },
            Object {
              "backgroundColor": "coral",
              "height": 40,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "borderColor": "#000",
                "borderRightWidth": 0,
                "borderTopWidth": 0,
              },
              Object {
                "justifyContent": "center",
              },
              undefined,
              undefined,
              Object {
                "flex": 1,
              },
              false,
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "white",
                  "textAlign": "center",
                },
                Object {
                  "backgroundColor": "transparent",
                },
              ]
            }
          >
            Rank
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "borderColor": "#000",
                "borderRightWidth": 0,
                "borderTopWidth": 0,
              },
              Object {
                "justifyContent": "center",
              },
              undefined,
              undefined,
              Object {
                "flex": 5,
              },
              false,
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "white",
                  "textAlign": "center",
                },
                Object {
                  "backgroundColor": "transparent",
                },
              ]
            }
          >
            Nickname
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "borderColor": "#000",
                "borderRightWidth": 0,
                "borderTopWidth": 0,
              },
              Object {
                "justifyContent": "center",
              },
              undefined,
              undefined,
              Object {
                "flex": 1,
              },
              false,
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "white",
                  "textAlign": "center",
                },
                Object {
                  "backgroundColor": "transparent",
                },
              ]
            }
          >
            Score
          </Text>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    bounces={false}
    scrollEnabled={true}
  >
    <View>
      <View
        style={
          Array [
            undefined,
            Object {
              "borderBottomWidth": 0,
              "borderColor": "#000",
              "borderLeftWidth": 0,
            },
          ]
        }
      >
        <View
          style={
            Array [
              undefined,
              0,
              Object {
                "flexDirection": "row",
                "overflow": "hidden",
              },
              Array [
                Object {
                  "backgroundColor": "oldlace",
                },
                false,
              ],
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                },
                Object {
                  "justifyContent": "center",
                },
                undefined,
                undefined,
                Object {
                  "flex": 1,
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Array [
                    Object {
                      "color": "coral",
                      "margin": 6,
                      "textAlign": "center",
                    },
                    false,
                  ],
                  Object {
                    "backgroundColor": "transparent",
                  },
                ]
              }
            >
              1
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                },
                Object {
                  "justifyContent": "center",
                },
                undefined,
                undefined,
                Object {
                  "flex": 5,
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Array [
                    Object {
                      "color": "coral",
                      "margin": 6,
                      "textAlign": "center",
                    },
                    false,
                  ],
                  Object {
                    "backgroundColor": "transparent",
                  },
                ]
              }
            >
              BigPurpleDonkey
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                },
                Object {
                  "justifyContent": "center",
                },
                undefined,
                undefined,
                Object {
                  "flex": 1,
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Array [
                    Object {
                      "color": "coral",
                      "margin": 6,
                      "textAlign": "center",
                    },
                    false,
                  ],
                  Object {
                    "backgroundColor": "transparent",
                  },
                ]
              }
            >
              3
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              undefined,
              0,
              Object {
                "flexDirection": "row",
                "overflow": "hidden",
              },
              Array [
                Object {
                  "backgroundColor": "oldlace",
                },
                false,
              ],
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                },
                Object {
                  "justifyContent": "center",
                },
                undefined,
                undefined,
                Object {
                  "flex": 1,
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Array [
                    Object {
                      "color": "coral",
                      "margin": 6,
                      "textAlign": "center",
                    },
                    false,
                  ],
                  Object {
                    "backgroundColor": "transparent",
                  },
                ]
              }
            >
              2
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                },
                Object {
                  "justifyContent": "center",
                },
                undefined,
                undefined,
                Object {
                  "flex": 5,
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Array [
                    Object {
                      "color": "coral",
                      "margin": 6,
                      "textAlign": "center",
                    },
                    false,
                  ],
                  Object {
                    "backgroundColor": "transparent",
                  },
                ]
              }
            >
              BigRedDonkey
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                },
                Object {
                  "justifyContent": "center",
                },
                undefined,
                undefined,
                Object {
                  "flex": 1,
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Array [
                    Object {
                      "color": "coral",
                      "margin": 6,
                      "textAlign": "center",
                    },
                    false,
                  ],
                  Object {
                    "backgroundColor": "transparent",
                  },
                ]
              }
            >
              2
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              undefined,
              0,
              Object {
                "flexDirection": "row",
                "overflow": "hidden",
              },
              Array [
                Object {
                  "backgroundColor": "oldlace",
                },
                false,
              ],
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                },
                Object {
                  "justifyContent": "center",
                },
                undefined,
                undefined,
                Object {
                  "flex": 1,
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Array [
                    Object {
                      "color": "coral",
                      "margin": 6,
                      "textAlign": "center",
                    },
                    false,
                  ],
                  Object {
                    "backgroundColor": "transparent",
                  },
                ]
              }
            >
              3
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                },
                Object {
                  "justifyContent": "center",
                },
                undefined,
                undefined,
                Object {
                  "flex": 5,
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Array [
                    Object {
                      "color": "coral",
                      "margin": 6,
                      "textAlign": "center",
                    },
                    false,
                  ],
                  Object {
                    "backgroundColor": "transparent",
                  },
                ]
              }
            >
              BigBlueDonkey
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderRightWidth": 0,
                  "borderTopWidth": 0,
                },
                Object {
                  "justifyContent": "center",
                },
                undefined,
                undefined,
                Object {
                  "flex": 1,
                },
                false,
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Array [
                    Object {
                      "color": "coral",
                      "margin": 6,
                      "textAlign": "center",
                    },
                    false,
                  ],
                  Object {
                    "backgroundColor": "transparent",
                  },
                ]
              }
            >
              1
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
